### ISSUE TITLE
[ML] Add new loss functions for supervised learning

### ISSUE THREAD

**Issue by valeriy42** [elastic/ml-cpp#973]
Many ML tasks require require optimization wrt a different objective than the least squares and cross-entropy defined already. Hence, for our supervised learning to be useful in the real world, we need to implement other loss functions.

- [x] The list of the "most useful" loss functions is still to be determined.
- [x] Check if we can substitute MAE for pseudo Huber loss

**Comment by valeriy42** [comment]
@Winterflower, @blaklaybul, @viglia what are the most important loss functions from your perspective?

@joshdevins Is there a specific loss function you need to use in the future, e.g. for learning to rank?

**Comment by valeriy42** [comment]
The current suggestions are:
* *[mean sqaured logarithimic error](https://scikit-learn.org/stable/modules/model_evaluation.html#mean-squared-logarithmic-error)* for dealing with exponentially growing targets (power law)
* *[(Pseudo) Huber loss](https://en.wikipedia.org/wiki/Huber_loss#Pseudo-Huber_loss_function)* for dealing with outliers

WDYT?

**Comment by joshdevins** [comment]
It's a longer story and there are a wide variety of approaches and loss functions, but I think I can summarize off the top of my head to get started (I can find more references if needed). Typically when comparing rankings we use the usual measures like nDCG, ERR, MAP, Kendall's Tau, etc. (in that order I guess). But these of course aren't convex so surrogate functions are used depending on the method. There are three approaches:

- Point-wise: Input is a document, output is a predicted score or class with "relevance" likelihood. Uses standard regression, classification or ordinal regression loss functions so hinge, logistic, zero-one loss, etc.
- Pair-wise: Input is a pair of documents, output is rank order. Uses RankNet, LambdaRank, LambdaMART, RankSVM, etc. I think RankSVM uses hinge-loss, LambdaRank is cross entropy. I have to look up the others [1]
- List-wise: Input is a list of documents, output is an ordered list. Uses ListNet (cross entropy), AdaRank tries to directly optimize nDCG

[1] [From RankNet to LambdaRank to
LambdaMART: An Overview](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/MSR-TR-2010-82.pdf)

**Comment by joshdevins** [comment]
tl;dr I guess hinge and log loss since we have cross entropy already 🤷‍♂  I'm honestly not sure if we would implement any of the LTR methods aside from the most basic or use something like RankLib or wait for 3rd party lib integration.

**Comment by blaklaybul** [comment]
I'm not too picky when it comes to loss functions - I will typically use generic ones provided by scikit-learn. More important to me is being able to publish our results to kaggle and more generally comparing our performance against others. A quick analysis of 2000+ kaggle competitions show the 20 most popular evaluation metrics

Evaluation Metric|Count
:-----:|:-----:
Categorization Accuracy|746
Root Mean Squared Error|409
Area Under Receiver Operating Characteristic Curve|325
Mean F-Score|133
Mean Absolute Error|123
Log Loss|49
Root Mean Squared Logarithmic Error|46
Mean Squared Error|37
Multiclass Loss|37
FScoreBeta|32
Levenshtein Mean|27
QuadraticWeightedKappa|10
Matthews correlation coefficient|8
Root Mean Square Percentage Error|7
Mean Columnwise Area Under Receiver Operating Characteristic Curve|7
Weighted Categorization Accuracy|7
Mean Average Precision at K|7
R^2 score (coefficient of determination)|7
Mean F-Score Beta|7
Mean Consequential Error|6

**Comment by valeriy42** [comment]
Cool, thank you @blaklaybul for the statistics. It seems that if we implement MAE and Mean Squared Logarithmic Error, we are in a very good shape!

**Comment by valeriy42** [comment]
> tl;dr I guess hinge and log loss since we have cross entropy already man_shrugging I'm honestly not sure if we would implement any of the LTR methods aside from the most basic or use something like RankLib or wait for 3rd party lib integration (depending on license reviews).

The problem with Hinge loss, is that it is not everywhere differentiable and has 0 second derivative. Since cross-entropy offers a close approximation of Hinge without the numerical drawback, we will probably try to get away with it.

**Comment by valeriy42** [comment]
Using MAE or Huber or Pseudo-Huber with delta set to 1.0 does not make any difference (see the diagrams below). Varying delta, we can mitigate the effect of outliers.

![Screenshot_2020-01-30 Untitled - Jupyter Notebook(1)](https://user-images.githubusercontent.com/1292899/73454343-2c7bdd00-436e-11ea-8056-2b2db5fe0986.png)

![Screenshot_2020-01-30 Untitled - Jupyter Notebook](https://user-images.githubusercontent.com/1292899/73453982-6e585380-436d-11ea-9b75-6314a65e4773.png)

Hence, I suggest that we implement pseudo Huber with delta set default to 1.0 instead of MAE.

WDYT @blaklaybul @tveasey ?

**Comment by stevedodson** [comment]
Here are the optimization objectives for Google AutoML:

Optimization objective | Problem type | API value | Use this objective if you want to...
-- | -- | -- | --
AUC ROC | Classification | MAXIMIZE_AU_ROC | Distinguish between classes. Default value for binary classification.
Log loss | Classification | MINIMIZE_LOG_LOSS | Keep prediction probabilities as accurate as possible. Default value for multi-class classification.
AUC PR | Classification | MAXIMIZE_AU_PRC | Optimize results for predictions for the less common class.
RMSE | Regression | MINIMIZE_RMSE | Capture more extreme values accurately (default).
MAE | Regression | MINIMIZE_MAE | View extreme values as outliers with less impact on model.
RMSLE | Regression | MINIMIZE_RMSLE | Penalize error on relative size rather than absolute value. Especially helpful when both predicted and actual values can be quite large.

**Comment by tveasey** [comment]
> Hence, I suggest that we implement pseudo Huber with delta set default to 1.0 instead of MAE.

I agree. Personally, I'd like to provide a dashboard which allows one to explore error characteristics efficiently as a function of the choice for delta. This would allow one to explore tradeoffs regarding the choice of MSE vs MAE in an interactive manner. As we index the hold out predictions, this ought to be easy with existing Elasticsearch functionality, i.e. run multiple times and then overlay, filter by feature values, etc prediction errors for different choices delta.

**Comment by stevedodson** [comment]
@valeriy42 @tveasey can you confirm which objectives you are intending to implement?

**Comment by tveasey** [comment]
We have log-loss. Our current thinking is we'll implement RMSLE and Huber loss, with which we can essentially replicate MAE,  for regression. Other classification metrics we'll support, if any, are TBD.
